dev/plugins/hu.elte.txtuml.xtxtuml/src/hu/elte/txtuml/xtxtuml/common/XtxtUMLExternalityHelper.xtend

```diff
@@ -0,0 +1,103 @@
+package hu.elte.txtuml.xtxtuml.common;
+
+import hu.elte.txtuml.xtxtuml.xtxtUML.TUAttribute
+import hu.elte.txtuml.xtxtuml.xtxtUML.TUConstructor
+import hu.elte.txtuml.xtxtuml.xtxtUML.TUExternality
+import hu.elte.txtuml.xtxtuml.xtxtUML.TUModifiers
+import hu.elte.txtuml.xtxtuml.xtxtUML.TUOperation
+import java.util.HashMap
+import java.util.Map
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.xtext.xbase.XBlockExpression
+
+import static hu.elte.txtuml.xtxtuml.xtxtUML.TUExternality.*
+
+/**
+ * This class tells of any XtxtUML language element whether it is non-external,
+ * external, or has an external body. For performance reasons, this class caches
+ * the calculated externality for each language element it was given and all
+ * their direct or non-direct containers.
+ * <br/>
+ * A language element is considered
+ * <ul>
+ * <li>to have an external body if it was declared as such,</li>
+ * <li>external if
+ * <ul>
+ * <li>it was declared as such, or</li>
+ * <li>one of its containers is external, or</li>
+ * <li>the element is the body of an operation or constructor which is marked to
+ * have an external body. (Note that because of the first rule, this results in all
+ * parts of the block also being external.)</li>
+ * </ul>
+ * </li>
+ * <li>non-external in all other cases.</li>
+ * </ul>
+ * Querying the externality of an {@link EObject} which is not an XtxtUML
+ * language element does not yield an error, the object is simply considered
+ * non-external (as obviously neither of the other two can be applied). 
+ */
+class XtxtUMLExternalityHelper {
+
+	val Map<EObject, TUExternality> externalityMap = new HashMap
+
+	def isNonExternal(EObject object) {
+		return object.externality === NON_EXTERNAL
```

Comment:
It is more efficient. (And the result of the two operators must be the same as we compare enum constants here.)